src/common/api/worker/rest/DefaultEntityRestCache.ts

```diff
@@ -371,55 +367,94 @@ export class DefaultEntityRestCache implements EntityRestCache {
 		listId: Id | null,
 		ids: Array<Id>,
 		ownerEncSessionKeyProvider?: OwnerEncSessionKeyProvider,
+		opts: EntityRestClientLoadOptions = {},
 	): Promise<Array<T>> {
+		const cachingBehavior = getCacheModeBehavior(opts.cacheMode)
 		const entitiesInCache: T[] = []
-		const idsToLoad: Id[] = []
-		for (let id of ids) {
-			const cachedEntity = await this.storage.get(typeRef, listId, id)
-			if (cachedEntity != null) {
-				entitiesInCache.push(cachedEntity)
-			} else {
-				idsToLoad.push(id)
+
+		let idsToLoad: Id[]
+		if (cachingBehavior.readsFromCache) {
+			idsToLoad = []
+			for (const id of ids) {
+				const cachedEntity = await this.storage.get(typeRef, listId, id)
+				if (cachedEntity != null) {
+					entitiesInCache.push(cachedEntity)
+				} else {
+					idsToLoad.push(id)
+				}
 			}
+		} else {
+			idsToLoad = ids
 		}
-		const entitiesFromServer: T[] = []
+
 		if (idsToLoad.length > 0) {
-			const entities = await this.entityRestClient.loadMultiple(typeRef, listId, idsToLoad, ownerEncSessionKeyProvider)
-			for (let entity of entities) {
-				await this.storage.put(entity)
-				entitiesFromServer.push(entity)
+			const entitiesFromServer = await this.entityRestClient.loadMultiple(typeRef, listId, idsToLoad, ownerEncSessionKeyProvider, opts)
+			if (cachingBehavior.writesToCache) {
+				for (const entity of entitiesFromServer) {
+					await this.storage.put(entity)
+				}
 			}
+			return entitiesFromServer.concat(entitiesInCache)
+		} else {
+			return entitiesInCache
 		}
-
-		return entitiesFromServer.concat(entitiesInCache)
 	}
 
-	async loadRange<T extends ListElementEntity>(typeRef: TypeRef<T>, listId: Id, start: Id, count: number, reverse: boolean): Promise<T[]> {
+	async loadRange<T extends ListElementEntity>(
+		typeRef: TypeRef<T>,
+		listId: Id,
+		start: Id,
+		count: number,
+		reverse: boolean,
+		opts: EntityRestClientLoadOptions = {},
+	): Promise<T[]> {
 		if (this.storage.getCustomCacheHandlerMap(this.entityRestClient).has(typeRef)) {
 			return await this.storage.getCustomCacheHandlerMap(this.entityRestClient).get(typeRef)!.loadRange(this.storage, listId, start, count, reverse)
 		}
 
 		const typeModel = await resolveTypeReference(typeRef)
-		if (!isCachedType(typeModel, typeRef)) {
-			return this.entityRestClient.loadRange(typeRef, listId, start, count, reverse)
+		const useCache = await this.shouldUseCache(typeRef, opts, typeModel)
+		if (!useCache) {
+			return await this.entityRestClient.loadRange(typeRef, listId, start, count, reverse, opts)
+		}
+
+		const behavior = getCacheModeBehavior(opts.cacheMode)
+		if (!behavior.readsFromCache) {
+			throw new ProgrammingError("cannot write to cache without reading with range requests")
```

Comment:
I suggest we do one of the following:
 - we remove it from `LoadOptions` for `loadRange` (so, a separate `LoadOptions` type)
 - we assume that not reading from cache also means not writing to cache in the case of range requests
 - we have a different option for loading ranges, either using cache or not using cache with no in-betweens (which is what we want from range requests most of the time)

I'm not sure what is better, the second one *might* be useful (especially for disjoined ranges) but it's also might be a surprising behavior.

Reply:
Actually let's leave it like it is for now. Sorry. We might implement it later.